wrappers/wrapper.rb

```diff
@@ -572,6 +573,88 @@ def detect_unfeasible_services(vrp)
+            to_delete_services = relation.linked_services
+            break false
+          else
+            to_delete_services << service
+          end
+        end
+      }
+
+      relation.linked_services -= to_delete_services
+
+      to_delete_services.each{ |service|
+        add_unassigned(unfeasible, vrp, service, "Service belongs to a relation of type #{relation.type} which makes it infeasible")
+      }
+      relation.linked_services.empty?
+    end
+
+    def last_service_reachable_sequence(vrp, vehicle, services)
+      vehicle_timewindow = vehicle.timewindow || Models::Timewindow.new(start: 0)
+      vehicle_duration = [vehicle.duration, vehicle_timewindow.safe_end(vehicle.cost_late_multiplier&.positive?) - vehicle_timewindow.start].compact.min
+      successive_activities = services.map{ |service|
+        [service.id, service.activity && [service.activity] || service.activities]
+      }
+      return nil if successive_activities.all?{ |_id, acts| acts.any?{ |a| a.timewindows.none?(&:end) } }
+
+      matrix = vrp.matrices.find{ |m| m.id == vehicle.matrix_id }
+      successive_activities = services.map{ |service|
+        [service.id, service.activity && [service.activity] || service.activities]
+      }
+
+      depot_approach = vehicle.start_point && matrix.time ? successive_activities.first.last.map{ |act| matrix.time[vehicle.start_point.matrix_index][act.point.matrix_index] }.min : 0
+      earliest_arrival = vehicle_timewindow.start + depot_approach
+
+      successive_activities.each.with_index{ |(id, a_activities), a_index|
+        return id if earliest_arrival > (a_activities.map{ |act| act.timewindows.map{ |tw| tw.safe_end(act.late_multiplier&.positive?) }.max }.max ||
+                                            vehicle_timewindow.safe_end(vehicle.cost_late_multiplier&.positive?))
+
+        last_duration = a_activities.map{ |a| a.duration_on(vehicle) }.min
+        depot_return = vehicle.end_point && matrix.time ? a_activities.map{ |act| matrix.time[act.point.matrix_index][vehicle.end_point.matrix_index] }.min : 0
+        earliest_depot_arrival = earliest_arrival + depot_return + last_duration
+
+        return id if earliest_depot_arrival > vehicle_timewindow.safe_end(vehicle.cost_late_multiplier&.positive?) ||
+                     earliest_depot_arrival - vehicle_timewindow.start > vehicle_duration
+
+        break if a_index == successive_activities.size - 1
+
+        _b_id, b_activities = successive_activities[a_index + 1]
+
+        earliest_arrival = a_activities.map{ |a_act|
+          b_activities.map{ |b_act|
+            travel_time = matrix.time && matrix.time[a_act.point.matrix_index][b_act.point.matrix_index]
+            travel_time += a_act.duration_on(vehicle) + travel_time > 0 ? a_act.setup_duration_on(vehicle) : 0
+            [b_act.timewindows.map(&:start)&.min || 0, earliest_arrival + travel_time].max
+          }.min
+        }.min
+      }
```

Comment:
it looks like we duplicated all the checks here in an elementary way. can't we generalize the original checks to prevent duplication?

Reply:
will be treated in another ticket